File: Trees/0310. Minimum Height Trees/0310.cpp
#pragma GCC optimize('0fast','unroll-loops')

class Solution {
public:
    unordered_map<int,vector<int>>adjList;
    pair<int,unordered_map<int,int>> bfs(int node) {
        deque<int>q;
        unordered_set<int>visited;
        q.push_back(node);
        visited.insert(node);
        unordered_map<int,int>parent;
        int farthestNode=node;

        while(!q.empty()){
            int curr=q.front();
            q.pop_front();
            for(auto& nei : adjList[curr]){
                if(visited.find(nei)==visited.end()){
                    visited.insert(nei);
                    q.push_back(nei);
                    farthestNode=nei;
                    parent[nei]=curr;
                }
            }
        }
        return {farthestNode,parent};
    }

    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        if(n==1){
            return {0};
        }

        for(auto& edge : edges){
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }

        pair<int,unordered_map<int,int>>p1=bfs(0);
        pair<int,unordered_map<int,int>>p2=bfs(p1.first);

        vector<int>path;
        int node=p2.first;
        while(node!=p1.first){
            path.push_back(node);
            node=(p2.second)[node];
        }
        path.push_back(p1.first);

        int mid=path.size()/2;
        if(path.size()%2==0){
            return {path[mid-1],path[mid]};
        } else{
            return {path[mid]};
        }
    }
};

File: Trees/0310. Minimum Height Trees/0310.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n==1:
            return [0]
        
        adjList=defaultdict(list)
        for u,v in edges:
            adjList[u].append(v)
            adjList[v].append(u)

        def bfs(node):
            q=deque([node])
            visited=set()
            visited.add(node)
            farthestNode=node
            parent=defaultdict(lambda:None)

            while(q):
                curr=q.popleft()
                for nei in adjList[curr]:
                    if nei not in visited:
                        visited.add(nei)
                        farthestNode=nei
                        parent[nei]=curr
                        q.append(nei)
            return farthestNode, parent
        
        a,_=bfs(0)
        b,parent=bfs(a)

        path=[]
        while(b is not None):
            path.append(b)
            b=parent[b]
        
        mid=len(path)//2
        if(len(path)%2==0):
            return [path[mid-1],path[mid]]
        else:
            return [path[mid]]